### PR TITLE
jobs: Start scheduled jobs daemon on server startup.

### PR DESCRIPTION
**Pull request by miretskiy**
Start scheduled job daemon when starting cockroach binary.

Informs #49346

Release notes: None

### CODE REVIEW

**Review by pbardea — COMMENTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[pkg/jobs/job_scheduler.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tljOq3JkvPi8S8Bdz:-M9tljOq3JkvPi8S8Be-:bsmw30m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler.go#L276)):*
> ```Go
> 		for waitPeriod := getInitialScanDelay(); ; waitPeriod = getWaitPeriod(sv) {
> 			select {
> 			case <-stopper.ShouldQuiesce():
> ```

Curious to learn more about this - what motivated this change? My (quite limited) understanding of the `Stopper` is that it supports running `Workers` and `Tasks`. When it is `Stop`ed it will first disallow running any more tasks, signal that the tasks should quiesce (via `ShouldQuiesce`), and then after all of the tasks have shut down will signal the workers that it's safe to shut down. I don't think changing this will cause any issue here - but I was wondering what motivated this change.
___
*[pkg/jobs/job_scheduler_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tlWix06I7UnugPqfY:-M9tlWiyCtuznpBCGK1E:b-v3kd3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler_test.go#L164)):*
> ```Go
> 
> 	// When disabled, we wait 5 minutes before rechecking.
> 	require.True(t, 5*time.Minute == getWaitPeriod(sv))
> ```

nit: Didn't notice this on my last review, but `require.Equal` may offer slightly clearer error messages on failure?
___
*[pkg/jobs/job_scheduler_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tlcnN-0YTUl2hg09T:-M9tlcnN-0YTUl2hg09U:be8yiwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler_test.go#L394)):*
> ```Go
> func TestJobSchedulerDaemonUsesSystemTables(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> 	defer settings.TestingSaveRegistry()
> ```

Should this be `settings.TestingSaveRegistry()()`?
___
*[pkg/jobs/job_scheduler_test.go, line 396 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tlaYeCS5-_3Q0xzpL:-M9tlaYeCS5-_3Q0xzpM:b-851u5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler_test.go#L396)):*
> ```Go
> 	defer settings.TestingSaveRegistry()
> 
> 	// Make daemon run quickly
> ```

nit: `.` at end of line

**Review by miretskiy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/jobs/job_scheduler.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tljOq3JkvPi8S8Bdz:-M9ty0385VECyytsz8MI:b-un32x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler.go#L276)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Curious to learn more about this - what motivated this change? My (quite limited) understanding of the `Stopper` is that it supports running `Workers` and `Tasks`. When it is `Stop`ed it will first disallow running any more tasks, signal that the tasks should quiesce (via `ShouldQuiesce`), and then after all of the tasks have shut down will signal the workers that it's safe to shut down. I don't think changing this will cause any issue here - but I was wondering what motivated this change.
</blockquote></details>

I was seeing errors messages in my test.  Basically, because I made scan loop super aggressive (10 ms), the stopper would first enter Quiesce() state -- by that time some things would be cancelled/quiesced, but I would still kick off work -- and that would result in an error.
So, basically, this would stop this worker as soon as stop has been called, as opposed to all tasks quiesced.
___
*[pkg/jobs/job_scheduler_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tlWix06I7UnugPqfY:-M9tyTwtBMPyYALd7tCp:bnlcj1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler_test.go#L164)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: Didn't notice this on my last review, but `require.Equal` may offer slightly clearer error messages on failure?
</blockquote></details>

Sorry, I didn't notice it last time either.
require.Equal() would not work (because of different types).  But require.EqualValues does work.  Fixed.
___
*[pkg/jobs/job_scheduler_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tlcnN-0YTUl2hg09T:-M9tyli79mGPoi2S2QcB:b-se0zw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler_test.go#L394)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Should this be `settings.TestingSaveRegistry()()`?
</blockquote></details>

Yes, thanks.  I really don't like Go here.

**Review by pbardea — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @pbardea)

___
*[pkg/jobs/job_scheduler.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50240#-M9tljOq3JkvPi8S8Bdz:-M9u0IoPEnsq03LcZukI:by4lbz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0d50be06545e0cd654f346d3deb100ee913243/pkg/jobs/job_scheduler.go#L276)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I was seeing errors messages in my test.  Basically, because I made scan loop super aggressive (10 ms), the stopper would first enter Quiesce() state -- by that time some things would be cancelled/quiesced, but I would still kick off work -- and that would result in an error.
So, basically, this would stop this worker as soon as stop has been called, as opposed to all tasks quiesced.
</blockquote></details>

Hmm, interesting. Thanks for the explanation.
